xmodule/library_content_block.py

```diff
@@ -104,6 +104,91 @@ class LegacyLibraryContentBlock(ItemBankMixin, XModuleToXBlockMixin, XBlock):
         scope=Scope.settings,
     )
 
+    def get_migrated_upstream_for_child(self, child_block_id: str) -> LibraryUsageLocatorV2:
+        """
+        Return the key of the V2 library block to which the indicated child's source block was migrated.
+
+        May raise UpstreamLink for a variety of reasons (@@TODO -- describe here).
+        """
+        # @@TODO Move these imports up to the top.
+        # @@TODO Factor the ContentLibraryMigration fetching into a nice Python API with data objects.
+        from openedx.core.djangoapps.content_libraries.models import ContentLibraryMigration
+        from cms.lib.xblock.upstream_sync import NoUpstream, BadUpstream
+        from xmodule.util.keys import BlockKey, derive_key
+        from opaque_keys import InvalidKeyError
+
+        if not self.source_library_id:
+            raise NoUpstream()
+        try:
+            source_library_key = self.source_library_key
+        except InvalidKeyError as exc:
+            # Source v1 library key is malformed.
+            # The legacy library content block UI will inform the user of this, so we don't need to.
+            # @@TODO Should we actually move error messaging here, so that we can remove that bit of the legacy UI?
+            raise NoUpstream() from exc
+        try:
+            library_migration = ContentLibraryMigration.objects.get(source_key=source_library_key)
+        except ContentLibraryMigration.DoesNotExist as exc:
+            # Source v1 library has not (yet?) been migrated to a v2 library.
+            # Alternatively, we are on an instance that doesn't have the source v1 library-- in that case, the legacy
+            # library content block UI will tell the user such.
+            # @@TODO Should we actually move error messaging here, so that we can remove that bit of the legacy UI?
+            raise NoUpstream() from exc
+
+        # In order identify the new v2 library block, we need to know the v1 library block that this child came from.
+        # Unfortunately, there's no straightforward mapping from these children back to their v1 library source blocks.
+        # (ModuleStore does have a get_original_usage function that inspects edit_info, but we can't count on
+        #  that always working, particularly if this block's course was imported from another instance.)
+        # However, we can work around this by just looping through every block in the legacy library, and testing to see
+        # if it's our source block.
```

Comment:
Can't we at least start with `get_block_original_usage` and fall back to this loop if it doesn't work?

Reply:
+1, but beware that at least one invocation of `get_block_original_usage` has cleanup code that implies it can pass back a versioned UsageKey:

https://github.com/openedx/edx-platform/blob/d19707793bacb44c9493844844f21ba3ca499e0e/xmodule/library_content_block.py#L408-L411